src/components/activities.js

```diff
@@ -0,0 +1,351 @@
+  }
+
+ /**
+  * Start an activity within the specified iframe.
+  * @param {!HTMLIFrameElement} iframe
+  * @param {string} url
+  * @param {?Object=} opt_args
+  * @return {!Promise<!ActivityIframePort>}
+  */
+  openIframe(iframe, url, opt_args) {
+    const activityPort = new ActivityIframePort(iframe, url, opt_args);
+    return activityPort.connect().then(() => activityPort);
+  }
+
+ /**
+  * Start an activity in a separate window. The result will be delivered
+  * to the `onResult` callback.
+  *
+  * The activity can be opened in two modes: "popup" and "redirect". This
+  * depends on the `target` value, but also on the browser/environment.
+  *
+  * The allowed `target` values are `_blank`, `_top` and name targets. The
+  * `_self`, `_parent` and similar targets are not allowed.
+  *
+  * The `_top` target indicates that the activity should be opened as a
+  * "redirect", while other targets indicate that the activity should be
+  * opened as a popup. The activity client will try to honor the requested
+  * target. However, it's not always possible. Some environments do not
+  * allow popups and they either force redirect or fail the window open
+  * request. In this case, the activity will try to fallback to the "redirect"
+  * mode.
+  *
+  * @param {string} requestId
+  * @param {string} url
+  * @param {string} target
+  * @param {?Object=} opt_args
+  * @param {?web-activities/activity-ports.ActivityOpenOptions=} opt_options
+  * @return {{targetWin: ?Window}}
+  */
+  open(requestId, url, target, opt_args, opt_options) {
+    return this.activityPorts_.open(
+        requestId, url, target, opt_args, opt_options);
+  }
+
+ /**
+  * Registers the callback for the result of the activity opened with the
+  * specified `requestId` (see the `open()` method). The callback is a
+  * function that takes a single `ActivityPort` argument. The client
+  * can use this object to verify the port using it's origin, verified and
+  * secure channel flags. Then the client can call
+  * `ActivityPort.acceptResult()` method to accept the result.
+  *
+  * The activity result is handled via a separate callback because of a
+  * possible redirect. So use of direct callbacks and/or promises is not
+  * possible in that case.
+  *
+  * A typical implementation would look like:
+  * ```
+  * ports.onResult('request1', function(port) {
+  *   port.acceptResult().then(function(result) {
+  *     // Only verified origins are allowed.
+  *     if (result.origin == expectedOrigin &&
+  *         result.originVerified &&
+  *         result.secureChannel) {
+  *       handleResultForRequest1(result);
+  *     }
+  *   });
+  * })
+  *
+  * ports.open('request1', request1Url, '_blank');
+  * ```
+  *
+  * @param {string} requestId
+  * @param {!function(!web-activities/activity-ports.ActivityResult):boolean} verifier
+  * @param {function(!Promise<!Object>)} callback
+  */
+  onResult(requestId, verifier, callback) {
```

Comment:
Question: would things be easier for us if we also were to start with two of these? `onResult` with verifier and the `Promise<Message>` and also `onResultDeprecated` with exactly the same old style.

Reply:
i can make it be onResultDeprecated(requestId, verifier, callback) for now. 
and then add onResult(requestId, verifier, messageType, callback)

Reply:
is that okay?